### PR TITLE
Implemented a prototype for config file support using figment

### PR DESCRIPTION
**Pull request by VJMReichenbach**
This adresses #39.

To be clear: This is more a proof of concept, than an actual contribution. I don't really know what I'm doing and my rust knowledge is ... limited.

I used [figment](https://docs.rs/figment/latest/figment/index.html) to read in a config file, but I ran into a couple issues:

Figment has support for clap. But the [documentation](https://docs.rs/figment/latest/figment/index.html#for-cli-application-authors) only shows an example, where the config file overwrites the cli, not the other way around.
I am used to cli tools that read configuration in the order `global config -> local config -> cli`, which I find the most intuitive and usable. But clap creates during parsing default values for every argument. And it [doesn't support empty values anymore](https://github.com/clap-rs/clap/issues/1740). So you can't just merge the figments one after another, like shown in the example, because every value would just be overwritten by the `Cli` struct, that was created by `Cli::parse()`.
I "merged" them manually, by just creating a default config `Cli` struct, and then overwriting the config created from the config files with every parameter that wasn't changed via the cli.
Obviously the long list of `if` blocks is really ugly and stupid, but I couldn't think of a better way to do it.
I feel like there should be a way to do this via [figment merge](https://docs.rs/figment/latest/figment/struct.Figment.html#method.merge), but I just couldn't figure it out. Maybe this is a limitation of `clap` `derive`?

**tl;dr**:
`clap` creates a struct with all values. But I only need the values that have changed from the default.

---

Other notes:
- `ProjectDirs::from()` needs [a qualifier and an organization](https://docs.rs/directories/latest/directories/struct.ProjectDirs.html#method.from). I had no idea what to write there
- My implementation currently doesn't check for local configs, but that should be fairly easy to implement. The question is: Should tex-fmt search parent folders? Should there be a `root` option in the config file, like in [editorconfig](https://editorconfig.org/)?
- Currently if you set e.g. `--keep` to true in the config file, you can't overwrite it via the cli, because there is no `--dont-keep` option

### CODE REVIEW

**Comment by WGUNDERWOOD** [comment]
Thanks for your help here! This is looking promising as a first draft. I want to think a little more about the correct precedence for different config files and cli arguments, so I'll try to do that this weekend.

File: src/main.rs [hunk 1]
```diff
@@ -43,7 +48,80 @@ const LINE_END: &str = "\n";
...
+        verbose: false,
+        stdin: false,
+        quiet: false,
+        trace: false,
+        files: Vec::<String>::new(),
+        tab: 2,
+        usetabs: false,
+        wrap: 80,
+        wrap_min: 70,
+    };
```
Comment (cdesaintguilhem) [inline src/main.rs:68]: Just a small suggestion: this could be condensed by implementing (or maybe just deriving) the `Default` trait for `Cli` and calling `default()`.
```suggestion
    // initilize default config
    let default_options = Cli::default();
```

**Comment by WGUNDERWOOD** [comment]
I have come up with the following proposal for precedence of config files and CLI arguments. The parsing will be done in two steps.

1. Search for a config file named `tex-fmt.toml`, in the following order (highest priority to lowest):
a) File path passed with `--config` on the command line
b) Config file in the directory where tex-fmt is executed
c) If tex-fmt is executed in a git repository, check every parent dir up to the git root
d) User configuration at `~/.config/tex-fmt/` or equivalent on Windows/MacOS
e) Fall back to the default tex-fmt settings
2. Finally, any specific CLI arguments supplied will be applied.

**Comment by VJMReichenbach** [comment]
I updated the draft a little, but I think i made things worse, lol.
I changed all values in `Cli` to `Option`s. 
Because of that, I created a new `Config` struct, which the replaces the old `Cli` struct.
I feel like `figment` should be able to merge the updated `Cli` created by `Cli::parse()` with the other options, but it panics, because `Cli` uses `Option`s now.
I have yet to figure out how to do that.

In the meantime I overwrite the config file arguments with the Cli ones with a lot of `unwrap_or`s.
That works, but is a probably annoying to maintain.

I assume you don't want seperate `Cli` and `Config` structs, since it makes it way more complicated to add new argument, so this is probably a dead end.
I'll try make figment work the way I want, using the options, but I am not sure if that is still a good idea.

---

Also: Why are you using a `i8` for `tab` and not an `u8`?

**Comment by WGUNDERWOOD** [comment]
Thanks for pointing out the `tab: i8` issue; I have now fixed this on the `develop` branch to read `tab: u8`.

**Comment by WGUNDERWOOD** [comment]
Perhaps you could take a look at the new implementation for configuration files available on the `config` branch? I think I have implemented all of the functionality I initially want, including sourcing config files from various locations, and merging arguments provided from config files/CLI/defaults.

**Comment by VJMReichenbach** [comment]
Looks good to me. I still think there should be an easier way to do this, since this is a fairly common problem, but I haven't seen a better solution yet.
Since this draft is not necessary anymore, I'll close it.
